source/octf/c/iotrace_plugin.cpp

```diff
@@ -0,0 +1,202 @@
+        std::atomic_uint64_t tracingRefCounter;
+    };
+
+public:
+    IOTracePluginC(octf_iotrace_plugin_context_t context,
+                   const std::string &pluginId,
+                   uint32_t queueCount)
+            : IOTracePlugin(pluginId, queueCount)
+            , m_context(context)
+            , m_queueContext(queueCount) {
+        for (auto &queueContext : m_queueContext) {
+            queueContext.tracingRefCounter = 0;
+            queueContext.traceStoping = true;
+        }
+    }
+
+    virtual ~IOTracePluginC() = default;
+
+    bool startTrace() override {
+        bool result = IOTracePlugin::startTrace();
+
+        if (result) {
+            m_context->tracing_active = true;
+            for (auto &queueContext : m_queueContext) {
+                queueContext.traceStoping = false;
+            }
+        }
+
+        return result;
+    }
+
+    bool stopTrace() override {
+        m_context->tracing_active = false;
+
+        for (auto &queueContext : m_queueContext) {
+            queueContext.traceStoping = true;
+        }
+
+        // Poll for all ongoing traces completion
+        while (isTraceOngoing()) {
+            std::this_thread::sleep_for(std::chrono::milliseconds(10));
+        }
+
+        return IOTracePlugin::stopTrace();
+    }
+
+    void push(uint32_t ioQueueId, const void *trace, size_t size) override {
+        if (ioQueueId > getTraceQueueCount()) {
+            ioQueueId %= getTraceQueueCount();
+        }
+
+        auto &queueContext = m_queueContext[ioQueueId];
+
+        queueContext.tracingRefCounter++;
+        if (queueContext.traceStoping) {
+            // Tracing stop was requested
+            queueContext.tracingRefCounter--;
+            return;
+        }
+        IOTracePlugin::push(ioQueueId, trace, size);
+        queueContext.tracingRefCounter--;
+    }
+
+private:
+    inline bool isTraceOngoing() const {
+        for (const auto &queueContext : m_queueContext) {
+            if (queueContext.tracingRefCounter) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+private:
+    octf_iotrace_plugin_context_t m_context;
+    std::vector<QueueContext> m_queueContext;
+};
+
+extern "C" int octf_iotrace_plugin_create(
+        const struct octf_iotrace_plugin_cnfg *cnfg,
+        octf_iotrace_plugin_context_t *_context) {
+    IOTracePluginC *plugin = NULL;
+    auto context = new struct octf_iotrace_plugin_context;
+
+    if (nullptr == context) {
+        return -ENOMEM;
+    }
+
+    try {
+        const char *prefix = "[OCTF]";
+        log::cout << log::prefix << prefix;
+        log::debug << log::prefix << prefix;
+        log::verbose << log::prefix << prefix;
+        log::cerr << log::prefix << prefix;
+        log::critical << log::prefix << prefix;
+
+        context->ref_sid = 0;
+        context->tracing_active = false;
+        plugin = new IOTracePluginC(context, cnfg->id, cnfg->io_queue_count);
```

Comment:
We check for success of context allocation but here we do not check for success of allocation